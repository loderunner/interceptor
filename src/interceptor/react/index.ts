export { Provider as InterceptProvider, useIntercepts } from './intercept'
