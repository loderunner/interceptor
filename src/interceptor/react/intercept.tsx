import * as React from 'react'
import { createContext, useContext, useState } from 'react'

import * as intercept from '../intercept'

interface InterceptContextType {
  intercepts: intercept.Intercept[]
  addIntercept: (pattern: string) => void
  removeIntercept: (inter: intercept.Intercept) => void
}

const InterceptContext = createContext<InterceptContextType>({
  intercepts: [],
  addIntercept: () => {},
  removeIntercept: () => {},
})

interface Props {
  children: React.ReactNode
}

export const Provider = ({ children }: Props) => {
  const [intercepts, setIntercepts] = useState([...intercept.intercepts])

  const addIntercept = (pattern: string) => {
    intercept.addIntercept({ pattern, enabled: true })
    setIntercepts([...intercept.intercepts])
  }
  const removeIntercept = (inter: intercept.Intercept) => {
    intercept.removeIntercept(inter)
    setIntercepts([...intercept.intercepts])
  }

  return (
    <InterceptContext.Provider
      value={{ intercepts, addIntercept, removeIntercept }}
    >
      {children}
    </InterceptContext.Provider>
  )
}

export const useIntercepts = () => useContext(InterceptContext)
