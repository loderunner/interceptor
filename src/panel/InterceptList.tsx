import * as React from 'react'
import { useCallback, useEffect, useMemo, useRef, useState } from 'react'

import { Clear as ClearIcon, Plus as PlusIcon } from '@/icons'
import { useIntercept, useIntercepts } from '@/interceptor/hooks'

import List from './components/List'
import { useSelection } from './selection'

import type { Intercept } from '@/interceptor'
import type { ChangeEvent, KeyboardEvent, SyntheticEvent } from 'react'

interface ItemProps {
  interceptId?: string
  onDelete: (inter: Intercept) => void
}

const Item = ({ interceptId, onDelete }: ItemProps) => {
  if (interceptId === undefined) {
    throw new Error('missing intercept id')
  }

  const { selection, setSelection, selectionType } = useSelection()
  const { intercept, updateIntercept } = useIntercept(interceptId)

  const onToggleEnabled = useCallback(() => {
    updateIntercept({ enabled: !intercept.enabled })
  }, [intercept])

  const onClickDelete = useCallback(
    (e: SyntheticEvent) => {
      e.stopPropagation()
      onDelete(intercept)
    },
    [intercept, onDelete]
  )

  const onSelect = useCallback(
    (e: SyntheticEvent) => {
      e.stopPropagation()
      setSelection({ ...intercept })
    },
    [intercept]
  )

  const patternEditRef = useRef<HTMLInputElement>(null)
  const patternLabelRef = useRef<HTMLSpanElement>(null)

  const [editing, setEditing] = useState(false)

  const PatternEdit = useCallback(() => {
    const [pattern, setPattern] = useState(intercept.pattern)
    const onChangePattern = useCallback(
      (e: ChangeEvent<HTMLInputElement>) => setPattern(e.target.value),
      []
    )
    const onConfirm = useCallback(() => {
      setEditing(false)
      updateIntercept({ pattern })
    }, [pattern])

    const onKeyDown = useCallback(
      (e: KeyboardEvent<HTMLInputElement>) => {
        switch (e.key) {
          case 'Enter':
            setEditing(false)
            updateIntercept({ pattern })
            break
          case 'Escape':
            setEditing(false)
            break
          default:
            return
        }
        e.stopPropagation()
      },
      [pattern]
    )

    useEffect(() => {
      patternEditRef.current?.focus()
      patternEditRef.current?.select()
    }, [])

    return (
      <input
        className="fixed"
        type="text"
        value={pattern}
        onChange={onChangePattern}
        onBlur={onConfirm}
        onKeyDown={onKeyDown}
        ref={patternEditRef}
      />
    )
  }, [intercept])

  const onDoubleClickPattern = useCallback((e: SyntheticEvent) => {
    e.stopPropagation()
    setEditing((e) => !e)
    patternEditRef.current?.classList.add(
      `left-[${patternLabelRef.current?.clientLeft}px]`,
      `top-[${patternLabelRef.current?.clientTop}px]`,
      `w-[${patternLabelRef.current?.clientWidth}px]`
    )
  }, [])

  const className = useMemo(() => {
    let className = 'flex w-full select-none justify-between p-1'
    if (selectionType !== 'intercept') {
      return className
    }
    const s = selection as Intercept
    if (s.id === intercept.id) {
      className += ' bg-blue-100'
    }
    return className
  }, [selection, intercept])

  return (
    <div className={className} role="listitem" onClick={onSelect}>
      <div className="flex flex-grow items-center space-x-1 overflow-hidden">
        <input
          type="checkbox"
          className="focus:ring-0"
          checked={intercept.enabled}
          onClick={(e) => e.stopPropagation()}
          onChange={onToggleEnabled}
        />
        <span
          className="w-full bg-red-100 empty:before:content-['\200b'] overflow-hidden text-ellipsis whitespace-nowrap"
          onDoubleClick={onDoubleClickPattern}
          ref={patternLabelRef}
        >
          {intercept.pattern}
        </span>
        {editing ? <PatternEdit /> : null}
      </div>
      <button
        className="self-stretch"
        title="Delete intercept"
        onClick={onClickDelete}
      >
        <ClearIcon className="h-full w-auto" />
      </button>
    </div>
  )
}

interface Props {
  className?: string
}

const InterceptList = ({ className }: Props) => {
  const { selection, setSelection, selectionType } = useSelection()
  const { intercepts, addIntercept, removeIntercept } = useIntercepts()

  const onDeleteIntercept = useCallback(
    (inter: Intercept) => {
      if (inter.id === undefined) {
        throw new Error('missing intercept id')
      }
      if (selectionType === 'intercept') {
        const s = selection as Intercept
        if (s.id === inter.id) {
          setSelection(null)
        }
      }
      removeIntercept(inter.id)
    },
    [selection]
  )

  const items = useMemo(
    () =>
      intercepts.map((inter) => (
        <Item
          key={inter.id}
          interceptId={inter.id}
          onDelete={onDeleteIntercept}
        />
      )),
    [intercepts, selection]
  )

  const onAddIntercept = useCallback(
    (e: SyntheticEvent) => {
      e.stopPropagation()
      const inter = addIntercept({ pattern: '', enabled: true })
      setSelection({ ...inter })
    },
    [intercepts]
  )

  const header = useMemo(
    () => (
      <div className="flex select-none justify-between bg-slate-100 p-1">
        <span className="font-bold">Intercepts</span>
        <button
          className="self-stretch"
          title="Add intercept"
          onClick={onAddIntercept}
        >
          <PlusIcon className="h-full w-auto" />
        </button>
      </div>
    ),
    []
  )

  return (
    <List
      id="intercept-list"
      className={className}
      header={header}
      items={items}
    ></List>
  )
}

export default InterceptList
