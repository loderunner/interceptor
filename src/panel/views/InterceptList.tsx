import * as React from 'react'

import List from './components/List'

interface Props {
  className?: string
}

const InterceptList = ({ className }: Props) => {
  return <List className={className} header="Intercepts"></List>
}

export default InterceptList
