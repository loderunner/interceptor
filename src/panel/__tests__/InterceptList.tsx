/* eslint-disable @typescript-eslint/ban-ts-comment */
import { cleanup, fireEvent, render, screen } from '@testing-library/react'
import * as React from 'react'
import '@testing-library/jest-dom'

import InterceptList from '../InterceptList'

describe('[InterceptList]', () => {
  afterEach(() => {
    cleanup()
    jest.resetAllMocks()
  })

  it('should match empty snapshot', () => {
    const { container } = render(<InterceptList />)
    expect(container).toMatchSnapshot()
  })

  it('should match snapshots after clicking buttons', () => {
    const { container, getByRole } = render(<InterceptList />)

    const addButton = getByRole('button', { name: 'Add intercept' })
    fireEvent.click(addButton)
    expect(container).toMatchSnapshot()

    const delButton = getByRole('button', { name: 'Delete intercept' })
    fireEvent.click(delButton)
    expect(container).toMatchSnapshot()
  })
})
