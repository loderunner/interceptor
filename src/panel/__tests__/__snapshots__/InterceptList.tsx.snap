// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[InterceptList] should match empty snapshot 1`] = `
<div>
  <div
    class="flex flex-col "
    id="intercept-list"
    role="list"
  >
    <div
      class="flex select-none justify-between bg-slate-100 p-1"
    >
      <span
        class="font-bold"
      >
        Intercepts
      </span>
      <button
        class="self-stretch"
        title="Add intercept"
      >
        <svg
          class="h-full w-auto"
        />
      </button>
    </div>
    <div
      class="flex flex-col overflow-y-auto"
    />
  </div>
</div>
`;

exports[`[InterceptList] should match snapshots after clicking buttons 1`] = `
<div>
  <div
    class="flex flex-col "
    id="intercept-list"
    role="list"
  >
    <div
      class="flex select-none justify-between bg-slate-100 p-1"
    >
      <span
        class="font-bold"
      >
        Intercepts
      </span>
      <button
        class="self-stretch"
        title="Add intercept"
      >
        <svg
          class="h-full w-auto"
        />
      </button>
    </div>
    <div
      class="flex flex-col overflow-y-auto"
    >
      <div
        class="flex w-full select-none justify-between p-1 bg-blue-100"
        role="listitem"
      >
        <div
          class="flex flex-grow items-center space-x-1 overflow-hidden"
        >
          <input
            checked=""
            class="focus:ring-0"
            type="checkbox"
          />
          <span
            class="w-full bg-red-100 empty:before:content-['\\\\200b'] overflow-hidden text-ellipsis whitespace-nowrap"
          />
        </div>
        <button
          class="self-stretch"
          title="Delete intercept"
        >
          <svg
            class="h-full w-auto"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`[InterceptList] should match snapshots after clicking buttons 2`] = `
<div>
  <div
    class="flex flex-col "
    id="intercept-list"
    role="list"
  >
    <div
      class="flex select-none justify-between bg-slate-100 p-1"
    >
      <span
        class="font-bold"
      >
        Intercepts
      </span>
      <button
        class="self-stretch"
        title="Add intercept"
      >
        <svg
          class="h-full w-auto"
        />
      </button>
    </div>
    <div
      class="flex flex-col overflow-y-auto"
    />
  </div>
</div>
`;

exports[`[InterceptList] should match snapshots after clicking items 1`] = `
<div>
  <div
    class="flex flex-col "
    id="intercept-list"
    role="list"
  >
    <div
      class="flex select-none justify-between bg-slate-100 p-1"
    >
      <span
        class="font-bold"
      >
        Intercepts
      </span>
      <button
        class="self-stretch"
        title="Add intercept"
      >
        <svg
          class="h-full w-auto"
        />
      </button>
    </div>
    <div
      class="flex flex-col overflow-y-auto"
    >
      <div
        class="flex w-full select-none justify-between p-1"
        role="listitem"
      >
        <div
          class="flex flex-grow items-center space-x-1 overflow-hidden"
        >
          <input
            checked=""
            class="focus:ring-0"
            type="checkbox"
          />
          <span
            class="w-full bg-red-100 empty:before:content-['\\\\200b'] overflow-hidden text-ellipsis whitespace-nowrap"
          />
        </div>
        <button
          class="self-stretch"
          title="Delete intercept"
        >
          <svg
            class="h-full w-auto"
          />
        </button>
      </div>
      <div
        class="flex w-full select-none justify-between p-1 bg-blue-100"
        role="listitem"
      >
        <div
          class="flex flex-grow items-center space-x-1 overflow-hidden"
        >
          <input
            checked=""
            class="focus:ring-0"
            type="checkbox"
          />
          <span
            class="w-full bg-red-100 empty:before:content-['\\\\200b'] overflow-hidden text-ellipsis whitespace-nowrap"
          />
        </div>
        <button
          class="self-stretch"
          title="Delete intercept"
        >
          <svg
            class="h-full w-auto"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`[InterceptList] should match snapshots after clicking items 2`] = `
<div>
  <div
    class="flex flex-col "
    id="intercept-list"
    role="list"
  >
    <div
      class="flex select-none justify-between bg-slate-100 p-1"
    >
      <span
        class="font-bold"
      >
        Intercepts
      </span>
      <button
        class="self-stretch"
        title="Add intercept"
      >
        <svg
          class="h-full w-auto"
        />
      </button>
    </div>
    <div
      class="flex flex-col overflow-y-auto"
    >
      <div
        class="flex w-full select-none justify-between p-1 bg-blue-100"
        role="listitem"
      >
        <div
          class="flex flex-grow items-center space-x-1 overflow-hidden"
        >
          <input
            checked=""
            class="focus:ring-0"
            type="checkbox"
          />
          <span
            class="w-full bg-red-100 empty:before:content-['\\\\200b'] overflow-hidden text-ellipsis whitespace-nowrap"
          />
        </div>
        <button
          class="self-stretch"
          title="Delete intercept"
        >
          <svg
            class="h-full w-auto"
          />
        </button>
      </div>
      <div
        class="flex w-full select-none justify-between p-1"
        role="listitem"
      >
        <div
          class="flex flex-grow items-center space-x-1 overflow-hidden"
        >
          <input
            checked=""
            class="focus:ring-0"
            type="checkbox"
          />
          <span
            class="w-full bg-red-100 empty:before:content-['\\\\200b'] overflow-hidden text-ellipsis whitespace-nowrap"
          />
        </div>
        <button
          class="self-stretch"
          title="Delete intercept"
        >
          <svg
            class="h-full w-auto"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`[InterceptList] should match snapshots after clicking items 3`] = `
<div>
  <div
    class="flex flex-col "
    id="intercept-list"
    role="list"
  >
    <div
      class="flex select-none justify-between bg-slate-100 p-1"
    >
      <span
        class="font-bold"
      >
        Intercepts
      </span>
      <button
        class="self-stretch"
        title="Add intercept"
      >
        <svg
          class="h-full w-auto"
        />
      </button>
    </div>
    <div
      class="flex flex-col overflow-y-auto"
    >
      <div
        class="flex w-full select-none justify-between p-1"
        role="listitem"
      >
        <div
          class="flex flex-grow items-center space-x-1 overflow-hidden"
        >
          <input
            checked=""
            class="focus:ring-0"
            type="checkbox"
          />
          <span
            class="w-full bg-red-100 empty:before:content-['\\\\200b'] overflow-hidden text-ellipsis whitespace-nowrap"
          />
        </div>
        <button
          class="self-stretch"
          title="Delete intercept"
        >
          <svg
            class="h-full w-auto"
          />
        </button>
      </div>
      <div
        class="flex w-full select-none justify-between p-1 bg-blue-100"
        role="listitem"
      >
        <div
          class="flex flex-grow items-center space-x-1 overflow-hidden"
        >
          <input
            checked=""
            class="focus:ring-0"
            type="checkbox"
          />
          <span
            class="w-full bg-red-100 empty:before:content-['\\\\200b'] overflow-hidden text-ellipsis whitespace-nowrap"
          />
        </div>
        <button
          class="self-stretch"
          title="Delete intercept"
        >
          <svg
            class="h-full w-auto"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
